src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -294,47 +295,78 @@ private void CheckReadArguments(byte[] buffer, int offset, int count)
 
             // OK, see first if it's a multi-block transform and we can speed up things
-            if (bytesToDeliver > _outputBlockSize)
+            int blocksToProcess = bytesToDeliver / _outputBlockSize;
+
+            if (blocksToProcess > 1 && _transform.CanTransformMultipleBlocks)
             {
-                if (_transform.CanTransformMultipleBlocks)
+                byte[] tempInputBuffer = null;
+                byte[] tempOutputBuffer = null;
+
+                try
                 {
-                    int BlocksToProcess = bytesToDeliver / _outputBlockSize;
-                    int numWholeBlocksInBytes = BlocksToProcess * _inputBlockSize;
-                    byte[] tempInputBuffer = new byte[numWholeBlocksInBytes];
-                    // get first the block already read
-                    Buffer.BlockCopy(_inputBuffer, 0, tempInputBuffer, 0, _inputBufferIndex);
-                    amountRead = _inputBufferIndex;
-                    amountRead += useAsync ?
+                    int numWholeBlocksInBytes = blocksToProcess * _inputBlockSize;
+                    tempInputBuffer = ArrayPool<byte>.Shared.Rent(numWholeBlocksInBytes);
+
+                    amountRead = useAsync ?
                         await _stream.ReadAsync(new Memory<byte>(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex), cancellationToken) :
                         _stream.Read(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex);
 
-                    _inputBufferIndex = 0;
-                    if (amountRead <= _inputBlockSize)
+                    int totalInput = _inputBufferIndex + amountRead;
+
+                    // If there's still less than a block, copy the new data into the hold buffer and move to the slow read.
+                    if (totalInput < _inputBlockSize)
                     {
-                        _inputBuffer = tempInputBuffer;
-                        _inputBufferIndex = amountRead;
-                        goto slow;
+                        Buffer.BlockCopy(tempInputBuffer, _inputBufferIndex, _inputBuffer, _inputBufferIndex, amountRead);
+                        _inputBufferIndex = totalInput;
                     }
-                    // Make amountRead an integral multiple of _InputBlockSize
-                    int numWholeReadBlocksInBytes = (amountRead / _inputBlockSize) * _inputBlockSize;
-                    int numIgnoredBytes = amountRead - numWholeReadBlocksInBytes;
-                    if (numIgnoredBytes != 0)
+                    else
                     {
-                        _inputBufferIndex = numIgnoredBytes;
-                        Buffer.BlockCopy(tempInputBuffer, numWholeReadBlocksInBytes, _inputBuffer, 0, numIgnoredBytes);
+                        // Copy any held data into tempInputBuffer now that we know we're proceeding
+                        Buffer.BlockCopy(_inputBuffer, 0, tempInputBuffer, 0, _inputBufferIndex);
```

Comment:
I'm a little confused at the logic.  Why don't we need to do this copy in the other branch as well?

Reply:
Okay.  I am now convinced that a) this line doesn't matter (_inputBufferIndex will always be 0 here) and b) that I'm going to leave the line, since it's easy for it to matter in the future

The Read(Async) call says the maximum it's willing to accept is the amount that will put it up to exactly the number of bytes that would fill the maximum number of blocks the output can accept in one shot.

If it returns the requested amount
{
    numIgnoredBytes is 0, and _inputBufferIndex becomes 0.

    If the output buffer was block aligned then bytesToDeliver is 0 and we return.  If not, we enter "slow".
}
else
{
    numIgnoredBytes might not be 0, _inputBufferIndex might not be 0, but we're definitely going to "slow".
}

in slow we read until we have a full input block.  Then we transform, and turn _inputBufferIndex to 0.  We might loop.

If we call read in slow and get 0, we move to process final.  From process final we can never re-enter read (past the early "still return 0").

So every return path (that can come back) sets _inputBufferIndex to 0.  But if we add a loop anywhere, or let it return without going into "slow" (and break everyone like SslStream did :smile:) then it could matter.

Took a lot of test variations with that line commented out to convince myself.